File: L4MicrosoftEasy/ValidPalindrome.java
// public class ValidPalindrome {
//     public boolean isValidPalindrome(String s) {
//         int start = 0, last = s.length() - 1;
//         while (start <= last) {
//             if (s.charAt(start) != s.charAt(last)) {
//                 return false;
//             }
//             start++;
//             last--;
//         }
//         return true;
//     }
    
//     public static void main(String[] args) {
//         ValidPalindrome vp = new ValidPalindrome();
//         String s = "abcba";
//         System.out.println(vp.isValidPalindrome(s)); // Output: true
//     }
// }


public class ValidPalindrome {

    public boolean isValidPalindrome(String s) {
        int start = 0, last = s.length() - 1;
        while (start <= last) {
            // Skip non-alphanumeric characters
            while (start < last && !Character.isLetterOrDigit(s.charAt(start))) {
                start++;
            }
            while (start < last && !Character.isLetterOrDigit(s.charAt(last))) {
                last--;
            }
            // Compare characters (case-insensitive)
            if (Character.toLowerCase(s.charAt(start)) != Character.toLowerCase(s.charAt(last))) {
                return false;
            }
            start++;
            last--;
        }
        return true;
    }

    public static void main(String[] args) {
        ValidPalindrome vp = new ValidPalindrome();
        String s = "A man, a plan, a canal: Panama";
        System.out.println(vp.isValidPalindrome(s)); // Output: true
    }
}

File: L4MicrosoftEasy/ValidPalindrome.cpp
// #include <iostream>
// #include <string>

// class ValidPalindrome
// {
// public:
//     bool isValidPalindrome(const std::string &s)
//     {
//         int start = 0, last = s.length() - 1;
//         while (start <= last)
//         {
//             if (s[start] != s[last])
//             {
//                 return false;
//             }
//             start++;
//             last--;
//         }
//         return true;
//     }
// };

// int main()
// {
//     ValidPalindrome vp;
//     std::string s = "A man, a plan, a canal: Panama";
//     std::cout << std::boolalpha << vp.isValidPalindrome(s) << std::endl; 
//     return 0;
// }




#include <iostream>
#include <string>
#include <cctype>

class ValidPalindrome
{
public:
    bool isValidPalindrome(const std::string &s)
    {
        int start = 0, last = s.length() - 1;
        while (start <= last)
        {
            // Skip non-alphanumeric characters
            while (start < last && !std::isalnum(s[start]))
            {
                start++;
            }
            while (start < last && !std::isalnum(s[last]))
            {
                last--;
            }
            // Compare characters (case-insensitive)
            if (std::tolower(s[start]) != std::tolower(s[last]))
            {
                return false;
            }
            start++;
            last--;
        }
        return true;
    }
};

int main()
{
    ValidPalindrome vp;
    std::string s = "A man, a plan, a canal: Panama";
    std::cout << std::boolalpha << vp.isValidPalindrome(s) << std::endl; // Output: true
    return 0;
}
